Remove support for specializing ToString outside the standard library

bjorn3 opened the pull request:
This is the only trait specializable outside of the standard library. Before stabilizing specialization we will probably want to remove support for this. It was originally made specializable to allow a more efficient ToString in libproc_macro back when this way the only way to get any data out of a TokenStream. We now support getting individual tokens, so proc macros no longer need to call it as often.

bjorn3 committed:
Remove support for specializing ToString outside the standard library

This is the only trait specializable outside of the standard library.
Before stabilizing specialization we will probably want to remove
support for this. It was originally made specializable to allow a more
efficient ToString in libproc_macro back when this way the only way to
get any data out of a TokenStream. We now support getting individual
tokens, so proc macros no longer need to call it as often.

bjorn3 commented on library/proc_macro/src/lib.rs:
```diff
@@ -19,9 +19,6 @@
 #![feature(rustdoc_internals)]
-// This library is copied into rust-analyzer to allow loading rustc compiled proc macros.
-// Please avoid unstable features where possible to minimize the amount of changes necessary
-// to make it compile with rust-analyzer on stable.
```
This hasn't been the case anymore for years now.

petrochenkov commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit b805f305425d06337cc1b84653c2bfe7b831a772 with merge f557de4c3dacd27cfe1deb8a6286a175e03ea4b8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b805f305425d06337cc1b84653c2bfe7b831a772","merge_sha":"f557de4c3dacd27cfe1deb8a6286a175e03ea4b8"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12316863259/job/34378077926) [(plain)](https://github.com/rust-lang/rust/commit/b805f305425d06337cc1b84653c2bfe7b831a772/checks/34378077926/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.7s done
#16 DONE 41.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [ui] tests/ui/specialization/min_specialization/issue-79224.rs stdout ----
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/specialization/min_specialization/issue-79224/issue-79224.stderr"
diff of stderr:

+ error[E0520]: `to_string` specializes an item from a parent `impl`, but that item is not marked `default`
+    |
+ LL |     fn to_string(&self) -> String {
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    |
+    |
+    = note: parent implementation is in crate `alloc`
+ 
1 error[E0277]: the trait bound `B: Clone` is not satisfied
2   --> $DIR/issue-79224.rs:18:29
3    |

52 LL | impl<B: ?Sized + std::clone::Clone> Display for Cow<'_, B> {
54 
- error: aborting due to 4 previous errors
+ error: aborting due to 5 previous errors
56 
56 
- For more information about this error, try `rustc --explain E0277`.
+ Some errors have detailed explanations: E0277, E0520.
+ For more information about an error, try `rustc --explain E0277`.
58 

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/specialization/min_specialization/issue-79224.rs:13:5
+ error[E0520]: `to_string` specializes an item from a parent `impl`, but that item is not marked `default`
+    |
+ LL |     fn to_string(&self) -> String {
+    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    |
---
To only update this specific test, also pass `--test-args specialization/min_specialization/issue-79224.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/specialization/min_specialization/issue-79224.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/specialization/min_specialization/issue-79224" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0520]: `to_string` specializes an item from a parent `impl`, but that item is not marked `default`
   |
LL |     fn to_string(&self) -> String {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   |
   = note: parent implementation is in crate `alloc`

error[E0277]: the trait bound `B: Clone` is not satisfied
##[error]  --> /checkout/tests/ui/specialization/min_specialization/issue-79224.rs:18:29
   |
LL | impl<B: ?Sized> Display for Cow<'_, B> {
   |                             ^^^^^^^^^^ the trait `Clone` is not implemented for `B`
   = note: required for `B` to implement `ToOwned`
help: consider further restricting type parameter `B` with trait `Clone`
   |
   |
LL | impl<B: ?Sized + std::clone::Clone> Display for Cow<'_, B> {

error[E0277]: the trait bound `B: Clone` is not satisfied
##[error]  --> /checkout/tests/ui/specialization/min_specialization/issue-79224.rs:20:5
   |
   |
LL |     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `B`
   = note: required for `B` to implement `ToOwned`
help: consider further restricting type parameter `B` with trait `Clone`
   |
   |
LL | impl<B: ?Sized + std::clone::Clone> Display for Cow<'_, B> {

error[E0277]: the trait bound `B: Clone` is not satisfied
##[error]  --> /checkout/tests/ui/specialization/min_specialization/issue-79224.rs:20:13
   |
   |
LL |     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
   |             ^^^^ the trait `Clone` is not implemented for `B`
   = note: required for `B` to implement `ToOwned`
help: consider further restricting type parameter `B` with trait `Clone`
   |
   |
LL | impl<B: ?Sized + std::clone::Clone> Display for Cow<'_, B> {

error[E0277]: the trait bound `B: Clone` is not satisfied
##[error]  --> /checkout/tests/ui/specialization/min_specialization/issue-79224.rs:20:62
   |
   |
LL |       fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
LL | |         //~^ ERROR: the trait bound `B: Clone` is not satisfied [E0277]
LL | |         //~^ ERROR: the trait bound `B: Clone` is not satisfied [E0277]
LL | |         //~| ERROR: the trait bound `B: Clone` is not satisfied [E0277]
LL | |         //~| ERROR: the trait bound `B: Clone` is not satisfied [E0277]
LL | |         write!(f, "foo")
   | |_____^ the trait `Clone` is not implemented for `B`
   |
   = note: required for `B` to implement `ToOwned`
help: consider further restricting type parameter `B` with trait `Clone`
help: consider further restricting type parameter `B` with trait `Clone`
   |
LL | impl<B: ?Sized + std::clone::Clone> Display for Cow<'_, B> {

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0277, E0520.
```

</details>

bjorn3 committed:
Update test

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12317114845/job/34384268318)
Build commit: f557de4c3dacd27cfe1deb8a6286a175e03ea4b8 (`f557de4c3dacd27cfe1deb8a6286a175e03ea4b8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12317114845/job/34384268318"},"merge_sha":"f557de4c3dacd27cfe1deb8a6286a175e03ea4b8"} -->

lqd commented:
@rust-timer build f557de4c3dacd27cfe1deb8a6286a175e03ea4b8

rust-timer commented:
Queued f557de4c3dacd27cfe1deb8a6286a175e03ea4b8 with parent dd436ae2a628c523c967a7876873a96c44b1e382, future [comparison URL](https://perf.rust-lang.org/compare.html?start=dd436ae2a628c523c967a7876873a96c44b1e382&end=f557de4c3dacd27cfe1deb8a6286a175e03ea4b8).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.5 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12318796667/job/34384352768) [(plain)](https://github.com/rust-lang/rust/commit/f7b14035a4a6d8d6d2ba31c022f25964122beb0c/checks/34384352768/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/ui/ref_option/ref_option.rs (revision `all`) ... ok
tests/ui/ref_option/ref_option.all.fixed ... ok

FAILED TEST: tests/ui/to_string_trait_impl.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-8eb91c7b4ad6d561.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-36c2079d6b9e5c3d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-299b795bee1b90dc.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-53a04e15a41bd95c.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-49fd91c51cbf146b.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-a4b4cb2e17d4d2b3.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f899203c39fb4fd5.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-af229d1b866e602b.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-3fd8d2059b043c18.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-51e0b532a8691593.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-b4ff13f51c887e57.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-fe2d5f0030188a8a.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-fcb6c0205f5698f3.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--crate-type=lib" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/to_string_trait_impl.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/to_string_trait_impl.stderr` to the actual output
--- tests/ui/to_string_trait_impl.stderr
+++ <stderr output>
+++ <stderr output>
-error: direct implementation of `ToString`
+error[E0520]: `to_string` specializes an item from a parent `impl`, but that item is not marked `default`
+  --> tests/ui/to_string_trait_impl.rs:44:9
    |
-LL | / impl ToString for Point {
+LL |         fn to_string(&self) -> String {
+LL |         fn to_string(&self) -> String {
-LL | |     fn to_string(&self) -> String {
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
-   = help: prefer implementing `Display` instead
+   = note: parent implementation is in crate `alloc`
 
-error: direct implementation of `ToString`
+error[E0520]: `to_string` specializes an item from a parent `impl`, but that item is not marked `default`
+  --> tests/ui/to_string_trait_impl.rs:71:9
    |
    |
-LL | /     impl ToString for S<i32> {
-LL | |         fn to_string(&self) -> String {
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
-   = help: prefer implementing `Display` instead
```

</details>

rust-timer commented:
Finished benchmarking commit (f557de4c3dacd27cfe1deb8a6286a175e03ea4b8): [comparison URL](https://perf.rust-lang.org/compare.html?start=dd436ae2a628c523c967a7876873a96c44b1e382&end=f557de4c3dacd27cfe1deb8a6286a175e03ea4b8&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=dd436ae2a628c523c967a7876873a96c44b1e382&end=f557de4c3dacd27cfe1deb8a6286a175e03ea4b8&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.5%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-0.5%, 0.5%]  | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=dd436ae2a628c523c967a7876873a96c44b1e382&end=f557de4c3dacd27cfe1deb8a6286a175e03ea4b8&stat=max-rss)
<details>
<summary>Results (primary -3.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1%  | [2.1%, 2.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -4.0% | [-8.4%, -2.4%] | 6     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -3.2% | [-8.4%, 2.1%]  | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=dd436ae2a628c523c967a7876873a96c44b1e382&end=f557de4c3dacd27cfe1deb8a6286a175e03ea4b8&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=dd436ae2a628c523c967a7876873a96c44b1e382&end=f557de4c3dacd27cfe1deb8a6286a175e03ea4b8&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary -0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.1%]   | 26    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.0%] | 11    |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.3%, -0.1%] | 6     |
| All ❌✅ (primary)                 | 0.0%  | [-0.2%, 0.1%]  | 37    |
</details>

**Bootstrap**: 770.497s -> 767.005s (-0.45%)
**Artifact size**: 330.36 MiB -> 330.49 MiB (0.04%)

petrochenkov commented:
Perf is ok, but the CI is still failing.
@rustbot author

bjorn3 committed:
Stop handling specialization in clippy's to_string_trait_impl lint

ToString can no longer be specialized, so no need to account for it in
to_string_trait_impl either.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12330805381/job/34416678505) [(plain)](https://github.com/rust-lang/rust/commit/af94847e6be7bbabc68b372b1e0903dc954abb28/checks/34416678505/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/ui/ref_option/ref_option.rs (revision `all`) ... ok
tests/ui/ref_option/ref_option.all.fixed ... ok

FAILED TEST: tests/ui/to_string_trait_impl.rs
command: CLIPPY_CONF_DIR="tests" RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-8eb91c7b4ad6d561.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-36c2079d6b9e5c3d.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-299b795bee1b90dc.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-53a04e15a41bd95c.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-49fd91c51cbf146b.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-a4b4cb2e17d4d2b3.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-f899203c39fb4fd5.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-af229d1b866e602b.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-3fd8d2059b043c18.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-51e0b532a8691593.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libserde_derive-b4ff13f51c887e57.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-fe2d5f0030188a8a.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-fcb6c0205f5698f3.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--crate-type=lib" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui" "tests/ui/to_string_trait_impl.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/to_string_trait_impl.stderr` to the actual output
--- tests/ui/to_string_trait_impl.stderr
+++ <stderr output>
```

</details>

bjorn3 commented:
@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:hourglass: Testing commit 3f3f27bb84b46be8294e27c6a2d64f0c918f3679 with merge 4790a435cbcb55c94ccdef51bf7a9b2e55824528...
<!-- homu: {"type":"BuildStarted","head_sha":"3f3f27bb84b46be8294e27c6a2d64f0c918f3679","merge_sha":"4790a435cbcb55c94ccdef51bf7a9b2e55824528"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12335919293/job/34429865541)
Approved by: petrochenkov
Pushing 4790a435cbcb55c94ccdef51bf7a9b2e55824528 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12335919293/job/34429865541"},"merge_sha":"4790a435cbcb55c94ccdef51bf7a9b2e55824528"} -->

rust-timer commented:
Finished benchmarking commit (4790a435cbcb55c94ccdef51bf7a9b2e55824528): [comparison URL](https://perf.rust-lang.org/compare.html?start=7caf35b2e5401d7740fdc567fdc388059208150b&end=4790a435cbcb55c94ccdef51bf7a9b2e55824528&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=7caf35b2e5401d7740fdc567fdc388059208150b&end=4790a435cbcb55c94ccdef51bf7a9b2e55824528&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.5%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-0.4%, -0.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-0.4%, 0.5%]  | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7caf35b2e5401d7740fdc567fdc388059208150b&end=4790a435cbcb55c94ccdef51bf7a9b2e55824528&stat=max-rss)
<details>
<summary>Results (primary -2.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.6%  | [4.6%, 6.5%]    | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -               | 0     |
| Improvements ✅ <br /> (primary)   | -6.9% | [-12.8%, -2.7%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -               | 0     |
| All ❌✅ (primary)                 | -2.8% | [-12.8%, 6.5%]  | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=7caf35b2e5401d7740fdc567fdc388059208150b&end=4790a435cbcb55c94ccdef51bf7a9b2e55824528&stat=cycles:u)
<details>
<summary>Results (primary -4.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -4.9% | [-4.9%, -4.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -4.9% | [-4.9%, -4.9%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=7caf35b2e5401d7740fdc567fdc388059208150b&end=4790a435cbcb55c94ccdef51bf7a9b2e55824528&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary -0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1%  | [0.0%, 0.3%]   | 28    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.0%] | 10    |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.4%, -0.1%] | 6     |
| All ❌✅ (primary)                 | 0.0%  | [-0.2%, 0.3%]  | 38    |
</details>

**Bootstrap**: 770.572s -> 769.38s (-0.15%)
**Artifact size**: 333.16 MiB -> 333.20 MiB (0.01%)
